common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -211,63 +214,74 @@ public ManagedBuffer getMergedBlockData(String appId, int shuffleId, int reduceI
-    AppPathsInfo appPathsInfo = Preconditions.checkNotNull(appsPathInfo.get(appId),
+    AppAttemptPathsInfo appAttemptPathsInfo = Preconditions.checkNotNull(appsPathsInfo.get(appId),
       "application " + appId + " is not registered or NM was restarted.");
-    String[] activeLocalDirs = Preconditions.checkNotNull(appPathsInfo.activeLocalDirs,
+    String[] activeLocalDirs = Preconditions.checkNotNull(appAttemptPathsInfo.activeLocalDirs,
       "application " + appId
       + " active local dirs list has not been updated by any executor registration");
     return activeLocalDirs;
   }
 
-  @Override
-  public void applicationRemoved(String appId, boolean cleanupLocalDirs) {
-    logger.info("Application {} removed, cleanupLocalDirs = {}", appId, cleanupLocalDirs);
-    // TODO: [SPARK-33236] Change the message when this service is able to handle NM restart
-    AppPathsInfo appPathsInfo = Preconditions.checkNotNull(appsPathInfo.remove(appId),
-      "application " + appId + " is not registered or NM was restarted.");
-    Iterator<Map.Entry<AppShuffleId, Map<Integer, AppShufflePartitionInfo>>> iterator =
+  /**
+   * Clean up the AppShufflePartitionInfo for a specific application attempt.
+   * If attemptId is -1, it means to clean up all the AppShufflePartitionInfo from
+   * all the attempts. Otherwise, only the AppShufflePartitionInfo from the specific
+   * application attempt will be cleaned up.
+   */
+  private void cleanupShufflePartitionInfo(String appId, int attemptId) {
+    Iterator<Map.Entry<AppAttemptShuffleId, Map<Integer, AppAttemptShufflePartitionInfo>>> iterator =
       partitions.entrySet().iterator();
     while (iterator.hasNext()) {
-      Map.Entry<AppShuffleId, Map<Integer, AppShufflePartitionInfo>> entry = iterator.next();
-      AppShuffleId appShuffleId = entry.getKey();
-      if (appId.equals(appShuffleId.appId)) {
+      Map.Entry<AppAttemptShuffleId, Map<Integer, AppAttemptShufflePartitionInfo>> entry = iterator.next();
+      AppAttemptShuffleId appAttemptShuffleId = entry.getKey();
+      if (appId.equals(appAttemptShuffleId.appId)
+          && (attemptId == -1 || attemptId == appAttemptShuffleId.attemptId)) {
         iterator.remove();
-        for (AppShufflePartitionInfo partitionInfo : entry.getValue().values()) {
+        for (AppAttemptShufflePartitionInfo partitionInfo : entry.getValue().values()) {
           partitionInfo.closeAllFiles();
         }
       }
     }
+  }
+
+  @Override
+  public void applicationRemoved(String appId, boolean cleanupLocalDirs) {
+    logger.info("Application {} removed, cleanupLocalDirs = {}", appId, cleanupLocalDirs);
+    // TODO: [SPARK-33236] Change the message when this service is able to handle NM restart
+    AppAttemptPathsInfo appAttemptPathsInfo = Preconditions.checkNotNull(appsPathsInfo.remove(appId),
+      "application " + appId + " is not registered or NM was restarted.");
+    cleanupShufflePartitionInfo(appId, -1);
```

Comment:
There is a race between this cleanup and some executor adding an entry into partitions map.
In a nutshell, we need to maintain an `applications` `Set` which contains currently running apps.
a) In `registerExecutor`, add to this `Set`
b) In `applicationRemoved`, remove from this `Set` before doing `cleanupShufflePartitionInfo`.
c) In `getOrCreateAppShufflePartitionInfo`, in `partitions.computeIfAbsent`, check if application is in this `Set` before returning a valid `Map` - else return `null` (and handle a `null` value for `shufflePartitions`).

This handles the race condition of `getOrCreateAppShufflePartitionInfo` (from a remote executor) overlapping with `applicationRemoved` (on local NM).
This also means the race condition @otterc referenced [above](https://github.com/apache/spark/pull/32007/files#r638325094) will not occur (application removal will ensure all records are cleaned up - so stale entries might live utmost until application termination - not after).

Reply:
+CC @otterc - please do check if this handles the case you referenced to above.

Reply:
Thinking more, while the above minimizes the possibility of leak, it does not theoretically eliminate it.

For example:

If no entry in partitions map for shuffle:
a) T1: getOrCreateAppShufflePartitionInfo -> partitions.computeIfAbsent -> mappingFunction lambda
c) T1: `synchronized(applications) { if (applications.contains(appId))  { return id -> Maps.newConcurrentMap() } else reutrn null; }` -> returns new map.
d) Context switch to T2.
e) T2: applicationRemoved -> applications.remove(appId) -> finish iterating over partitions map.
f) T1: partitions completes adding to map.

Now we have an entry in partitions map which will never be cleaned up.
A solution to this would be something like:

```
Map<Integer, AppAttemptShufflePartitionInfo> shufflePartitions = partitions.get(appAttemptShuffleId);
if (null == shufflePartitions) {
  synchronized(applications) {
    if (applications.contains(appId)) {
      shufflePartitions = partitions.computeIfAbsent(appAttemptShuffleId, id -> Maps.newConcurrentMap());
    }
  }
}
```


This minimizes cost of synchronized to only when shuffle is not known for app attempt - while ensuring correctness.

Reply:
Btw, we could eliminate need for `applications` and use `appsPathsInfo` instead (and use partitions.compute in the if block above).

Reply:
Added as suggested.

Reply:
@mridulm FYI, I have not added the unit test for this memory leak yet in this update PR. Post the majority of the required change first and I will add it tomorrow.

Reply:
Will resolve the multi-attempts in SPARK-35546